src/analyses/warnPostProc.ml

```diff
@@ -0,0 +1,561 @@
+    | _, `Top -> x
+    | `Lifted _, `Lifted _ -> CondSet.fold (fun c acc -> union acc (union (tuples_of c x) (tuples_of c y))) conds (empty ())
+
+end
+
+module VarNode =
+struct
+  include Analyses.Var
+
+  let is_write_only _ = false
+  let node n = n
+
+  let pretty_trace = Node.pretty_trace
+
+end
+
+module VarNodeL =
+struct
+  include VarNode
+
+  let pretty_trace () x = Pretty.dprintf "In %a" pretty_trace x
+end
+
+module VarNodeG =
+struct
+  include VarNode
+
+  let pretty_trace () x = Pretty.dprintf "Out %a" pretty_trace x
+end
+
+module V =
+struct
+  include Printable.Std
+  include Constraints.Var2 (VarNodeL) (VarNodeG)
+
+  let pretty = pretty_trace
+
+  include Printable.SimplePretty (
+    struct
+      type nonrec t = t
+      let pretty = pretty
+    end
+    )
+end
+
+module HM = Hashtbl.Make (V)
+
+let hoistHM = HM.create 10
+
+let antSolHM = HM.create 10
+
+let sinkHM = HM.create 10
+
+let avSolHM = HM.create 10
+
+let ask : (Node0.t -> Queries.ask) ref = ref (fun a -> assert false)
+let ask' : (Node0.t -> (CilType.Location.t * Edge.t) -> Node0.t -> Queries.ask) ref = ref (fun n e n' -> assert false)
+
+let set_loc (message : Alarm.t) loc =
+  match message.multipiece with
+  | Single piece -> {message with multipiece = Single {piece with loc = Some loc}}
+  | Group group -> {message with multipiece = Group {group with loc = Some loc}}
+
+(* Do not add related alarm when an alarm with that location already exists in the pieces list
+   because otherwise the original alarm is set as related and duplicated *)
+let contains_alarm_with_loc node (pieces : M.Piece.t list) =
+  let node_equals loc node =
+    match loc with
+    | Some M.Location.Node n -> Node.equal n node
+    | _ -> false
+  in
+  List.fold (fun acc (piece : M.Piece.t) -> node_equals piece.loc node || acc) false pieces
+
+
+let set_related (message : Alarm.t) node =
+  match message.multipiece with
+  | Group group ->
+    if contains_alarm_with_loc node group.pieces then message else
+      let piece = RM.Piece.{text = "Possible cause"; loc = Some (RM.Location.Node node); context = None} in
+      let group = {group with pieces = (piece::group.pieces)} in
+      {message with multipiece = Group group}
+  | _ -> failwith "TODO"
+
+let filter_eq_locs loc1 loc2 =
+  match loc1, loc2 with
+  | Some (M.Location.Node n1), Some (M.Location.Node n2) -> not @@ Node.equal n1 n2
+  | _, _ -> true
+
+let rem_related (message : Alarm.t) node =
+  match message.multipiece with
+  | Group group ->
+    let pieces = List.filter (fun (piece : M.Piece.t) -> filter_eq_locs (Some (RM.Location.Node node)) piece.loc || piece.text != "Possible cause") group.pieces in (* TODO: remove the text check *)
+    let group = {group with pieces = pieces} in
+    {message with multipiece = Group group}
+  | _ -> failwith "TODO"
+
+let rec append_unique l1 l2 =
+  match l2 with
+  | [] -> l1
+  | (x::xs) -> if List.mem x l1
+    then append_unique l1 xs
+    else append_unique (x::l1) xs
```

Comment:
This is visible in some profiling with lots of `List.mem` calls. Probably would be better to have sets somewhere. Or maybe only deduplicate at the very end.

Also, `List.mem` does polymorphic equality which should be avoided.